Use python-gpg instead of pygpgme

dcbaker opened the pull request:
This is a request for comment, I think this is largely ready to land, but there are a few things I'm not exactly sure about in this, which are marked with XXX or FIXME. Currently this seems to work as well as the pygpgme implementation did, which is to say mostly. There is an unrelated bug that causes some signatures to fail to validate, though I haven't been able to figure out what it is that's causing that.

It is somewhat pressing to get this finished and merged, as debian is moving to deprecated pygpgme and remove it from their repositories.

lucc commented:
Two test fail for me locally: tests.crypto_test.TestListKeys.test_list_hint and test_list_no_hints. Both raise `GPGMEError: GPGME: End of file`.

The travis file will nee some love to make the tests run on travis. Currently both jobs fail because the gpg module is not istalled.

I will use this branch locally and do a full review as soon as possible.

dcbaker commented:
Interesting, they fail for me too when running with `python setup.py test`, but not with `pytest tests`. I'll look into it.

dcbaker commented:
@teythoon I dug into this and don't really understand. It seems like `op_keylist_next` on it's last iteration gets a GPGMEError that has code 107, and message "End of file reached". does that make any sense to you?

teythoon commented:
@dcbaker that is GPGMEs version of StopIteration, the message is right, though the code is off (107 can never be an error code, but an error source).

lucc commented:
I can not reply to an email where I don't have a public key for the recipient. I have encrypt_by_default=trusted in my config. Here is a part of the log:
```
DEBUG:globals:Trying to encrypt message because account.encrypt_by_default=trusted
DEBUG:utils:encryption keys: [u'XXX@XXX']
ERROR:ui:Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1532, in unwindGenerator
    return _inlineCallbacks(None, gen, Deferred())
  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File "/home/luc/vcs/alot/alot/commands/utils.py", line 44, in set_encrypt
    signed_only=signed_only)
  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1532, in unwindGenerator
    return _inlineCallbacks(None, gen, Deferred())
--- <exception caught here> ---
  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File "/home/luc/vcs/alot/alot/commands/utils.py", line 74, in _get_keys
    signed_only=signed_only)
  File "/home/luc/vcs/alot/alot/crypto.py", line 112, in get_key
    raise e
gpg.errors.KeyNotFound: XXX@XXX
```
This doesn't happen on master.

dcbaker commented:
@lucc ah, okay. That's just a matter of catching the exception I'll fix that.

lucc commented on alot/crypto.py:
```diff
@@ -81,53 +54,64 @@ def get_key(keyid, validate=False, encrypt=False, sign=False,
             # expired keys together with valid keys. If only one key is valid
             # for given operation maybe we can still return it instead of
             # raising exception
+            #
```
empty lines?

lucc commented on alot/crypto.py:
```diff
@@ -81,53 +54,64 @@ def get_key(keyid, validate=False, encrypt=False, sign=False,
+                            "(use a key ID like 4AC8EE1D).",
+                            code=GPGCode.AMBIGUOUS_NAME)
+                    valid_key = k
+            except gpg.errors.GPGMEError as e:
```
Which line of code in the try block is able to raise that?

dcbaker replied:
validate_key

dcbaker replied:
no. I don't know what i was thinking, not `validate_key`, in the generator returned by `list_key`

lucc replied:
You could also do
```py
try:
    for k in list_keys(hint=keyid):
        ...
```
then it would be even more obvious.

dcbaker replied:
okay, I can do that

lucc commented on alot/crypto.py:
```diff
@@ -81,53 +54,64 @@ def get_key(keyid, validate=False, encrypt=False, sign=False,
 
             if not valid_key:
-                # there were multiple keys found but none of them are valid for
+                # there were multiple keys found  but none of them are valid for
```
double space

lucc commented on alot/crypto.py:
```diff
@@ -196,17 +169,18 @@ def verify_detached(message, signature):
-    return status
+    ctx = gpg.core.Context()
+    try:
+        verify_results = ctx.verify(message, signature)[1]
```
Where csn I find the docs for gpg to check what this does? (Especially the `[1]`)

dcbaker replied:
look for `pyme` (which is what the gpg bindings were called before getting blessed as the official ones). http://pyme.sourceforge.net/doc/pyme/pyme.core.html

Honestly the documentation for the gpg bindings is sparse, and usually they just say "read what the C code in gpgme does", which is pretty frustrating. there are some docstrings, so pydoc can help too.

lucc replied:
I am sorry I understand neither the docs you linked to (and `verify` is not listed there) nor the code I installed via pip or in the code I cloned (it has the verify `method` but the second element in the returned tuple is created by `self.op_verify_result()` and I can not find the definition for this anywhere. I assume it was created by some metaprogramming and wraps the C function `gpgme_op_verify_result`).

lucc commented on alot/crypto.py:
```diff
@@ -215,45 +189,25 @@ def decrypt_verify(encrypted):
     :param encrypted: the mail to decrypt
     :returns: a tuple (sigs, plaintext) with sigs being a list of a
-              :class:`gpgme.Signature` and plaintext is a `str` holding
+              :class:`gpg.result.Signature` and plaintext is a `str` holding
```
grammer: "X is a list of a Y"?

dcbaker commented:
I'll see about fixing the whitespace and comment things and pushing a new version. I'm still not sure why those tests fail when run without pytest.

pazz commented:
I've tested this by sending myself an encrypted message, and something's not quite right.

log when sending:
```
DEBUG:ui:cmdline: 'send'
DEBUG:ui:envelope command string: "send"
DEBUG:__init__:mode:envelope got commandline "send"
DEBUG:__init__:ARGS: [u'send']
DEBUG:__init__:cmd parms {}
DEBUG:envelope:encrypting plaintext: Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: quoted-printable

this is an encrypted mail

DEBUG:envelope:<type 'str'>
```

I can open this as expected.
Now if I encrypt AND sign the message:
(log when opening the thread)
```
DEBUG:ui:cmdline: 'send'
DEBUG:ui:envelope command string: "send"
DEBUG:__init__:mode:envelope got commandline "send"
DEBUG:__init__:ARGS: [u'send']
DEBUG:__init__:cmd parms {}
DEBUG:envelope:signing plaintext: Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: quoted-printable

this should be signed and encrypted.

DEBUG:envelope:encrypting plaintext: Content-Type: multipart/signed; protocol="application/pgp-signature";
 micalg="pgp-sha1"; boundary="===============2742654288114383880=="
MIME-Version: 1.0
Content-Disposition: inline

--===============2742654288114383880==
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: quoted-printable

this should be signed and encrypted.

--===============2742654288114383880==
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Description: signature
Content-Type: application/pgp-signature; name="signature.asc"; charset="us-ascii"

-----BEGIN PGP SIGNATURE-----

iF0EABECAB0WIQS3g41l/ugN7R3KSU2UNANn19bFqgUCWXDD2gAKCRCUNANn19bF
qkrHAJ9y0P0/P6f3H2ky9DhHMYpSspTlFACgw1ZP/NqeAp7VSD06u/se1aFzILA=
=0M5w
-----END PGP SIGNATURE-----

--===============2742654288114383880==--

DEBUG:envelope:<type 'str'>
```


The mail "body" contains
```
Malformed OpenPGP message: Invocation of
gpgme_op_decrypt_verify
```

and the log says nothing about crypto stuff:
```
DEBUG:ui:search command string: "select"
DEBUG:__init__:mode:search got commandline "select"
DEBUG:__init__:ARGS: [u'select']
DEBUG:__init__:cmd parms {}
INFO:search:open thread view for thread:000000000000a8e8: test4
DEBUG:utils:unquoted header: |Patrick Totzke <patricktotzke@gmail.com>|
DEBUG:utils:unquoted header: |Patrick Totzke <patricktotzke@othermail.com>|
DEBUG:utils:unquoted header: |test4|
DEBUG:thread:want tags header
DEBUG:utils:unquoted header: |Thu, 20 Jul 2017 15:53:14 +0100|
DEBUG:utils:unquoted header: |<150056239492.31571.13277224763525518685@brick.local>|
DEBUG:buffers:Tbuffer: auto remove unread tag from msg?
DEBUG:buffers:Tbuffer: No, cursor on summary
DEBUG:buffers:Tbuffer: auto remove unread tag from msg?
DEBUG:buffers:Tbuffer: No, cursor on summary
```
The mail file contains this:
```
Return-Path: <patricktotzke@gmail.com>
Received: from localhost (dhcp-90-255.inf.ed.ac.uk. [129.215.90.255])
        by smtp.gmail.com with ESMTPSA id y100sm2647609wmh.7.2017.07.20.07.53.15
        for <patricktotzke@othermail.com>
        (version=TLS1_2 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256);
        Thu, 20 Jul 2017 07:53:15 -0700 (PDT)
Content-Type: multipart/encrypted; protocol="application/pgp-encrypted";
 boundary="===============0059215141468070322=="
MIME-Version: 1.0
To: Patrick Totzke <patricktotzke@othermail.com>
Message-ID: <150056239492.31571.13277224763525518685@brick.local>
From: Patrick Totzke <patricktotzke@gmail.com>
User-Agent: alot/0.6.0dev
Subject: test4
Date: Thu, 20 Jul 2017 15:53:14 +0100
X-TUID: Q347sUO3ix9B

--===============0059215141468070322==
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Type: application/pgp-encrypted; charset="us-ascii"

Version: 1
--===============0059215141468070322==
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Content-Type: application/octet-stream; charset="us-ascii"

-----BEGIN PGP MESSAGE-----

hQIMA2F/brkAvgbfAQ/+PLRRb/Z/5VVuaeDFSnrhfbOA3FsUp9GBZSEyagjzUjWs
trfUBb1Wo47AAIVpJljQgiVW3o7Y/lAYpd+hu+pRoMDkTlTZEZRIW1QIZjGD5X38
CzkpAUc50JMT/0mZzcIL09CP+QhFlOCL6t8ackTUGKfRX254RW79klu4yo2jUfrB
D0d9kO7ZQSKB0I/VHS2stWanJGo/c5H4YAMjdtrhXUY0qUgm/3kO0aZzkKwrtuat
dIBPgPnF5bGYcH4mjYfVs3RmifRFo08IislygiLVxjQRdXHp4RCuWeH9v4j2P1Oy
zrecrqmaRwBODYBeqOpVT5amz3ivAIHvgbiOwlerF2EYZl1p41pT1MSoK+JdMcWl
zbTEYooB921RQBaVk+kAG8SGgCZY8aWrEXDWrwnzXd8B+Xs/yn+mzyaL/zk411bB
u94UqWQHj8OKXeiF5TX7h7QveOl2FAEPVvQ2Nwj7U1vylujVMRnB6wIIuNfopn5Q
g6QHKu3zryyItCymYlkbaIdr8Jb7jJcVGbsKPnoB0nmHrsY8qCNLrhsd3vOYbmBN
Xxz681VzqCerBALNMJCZIzD3tRVDzHs7vMCv6CKDzNuPHThfWy/96sXnSAJwAATn
r7kitBC6JqebUq0GM4ilamYDSMUZHhZybx7ecx38/TCS2Uo7MQT0a8FJk2Nzat/S
6QHx3FP+UQLq5hNbPYWQm9/jZHlAtPYsGijcHoSIHcnYjHGKoHeeNAea1uQt2kj5
qHiDghSWjScwuzs8kL+XqOb2cTuFIqK3RygLj16G4Llfl6+in/vJWfXJwRWuA3LI
TtqsDA38+M1AyuEZMyL3qkAORzVekFUDBsr4+23LoX9U+oN0CEFJIcw1AdOIkdw4
irg9oVJgXKjgMmHejCa1d+3fhdFgSRJXn7SO0qLWr5y7CqqWRJVx4Rb/KMR4iqcX
DlC0AJMuAdo4CnqmSps7u73UwDaY7Rhwnl3t9zUM1adFoTlP3ntqiWfUwedoSZfD
IrBCc1OKFN6a9y73mpdm4VVlRmokso6y7RmyDvax8VRFYGGhLyvCqIbOZEiBv54v
uK63opvPFt5RC87K+EZbfraYuVbKXklNaaRrg9JgokYz5hf/Mngwr3sfgq0sfhFY
VvemDP3CVKObvAFia3iQnw2l5wNULpNuhG3JcjMHsV5uX5x37Vqqo8UbVXVyeN8J
LU2OuPoF/SGQS2yq6g2vHlugufkShctTMeyydNpqTltDaY6Rwyo612kQxCpnpQ44
3RbolnP5W8Bm6f/3Ew+lirgMzuesRv/c/HBCZQeVTDtheb28LNKtCJ8kz8uXu1iJ
q5Z/Y5sbG6uRev9xYrWDLvafVDhSFvyKeJL2RpiCDbxgfqLBWR/cfUqy8ju9w0ZW
BicM8ahlm1gapy2cw3tamtxxygsMV4T/GhN4I0DkkXEAyTD+30mzh+OO03vK388j
4sFNfB3cCKtWuJiIE3yirE8xeZMYM2Tmh7SrK/ZUzCZjiH0myc0YInkSzCM5MlWJ
vAT8ED39Ncoc/kt+yTTnIA==
=zP1+
-----END PGP MESSAGE-----

--===============0059215141468070322==--
```

pazz commented:
...and I've just noticed that the very same thing happens on the master branch, so this is probably only loosely related?

pazz commented:
Sorry for the spam :-) 
It seems that there is something wrong either with my gpg setup or the decryption part in alot: Ican read all those mails on my android tablet with k9.

dcbaker commented:
@pazz it wouldn't surprise me if we're doing something wrong with decryption, considering that until recently our signature validation code was completely broken...

lucc commented:
Ok I found the [general docs for gpgme](https://www.gnupg.org/documentation/manuals/gpgme/index.html). They do not mention anything about python but maybe they are still helpful for understanding the backend. It can also be build locally:
```sh
git clone git://git.gnupg.org/gpgme.git gpgme
cd gpgme
./autogen.sh
./configure --enable-maintainer-mode
make html
browser doc/gpgme.html/index.html
```

@teythoon did I miss some special python docs or is there any plan to make some language/binding specific docs? I know it sounds snobby but it feels clumsy to read the python source code and then figure out what part of the docs for the C code one has to read when coding in python (I might be spoiled by the python stdlib docs ;).

dcbaker commented:
I've updated this, I still haven't had a chance to look at lucc's encrypt_by_default bug, but I think I've addressed the rest of the comments. I'm working on the travis problems currently.

lucc commented on alot/crypto.py:
```diff
@@ -81,69 +55,93 @@ def get_key(keyid, validate=False, encrypt=False, sign=False,
 
+    The generator may raise exceptions of :class:gpg.errors.GPGMEError, and it
+    is the caller's responsibility to handle them.
+
```
just asthetics: In other places we used the `:raises: ...` syntax for that (like `:param ...:` or `:returns: ...`).

dcbaker replied:
I think that `:raises:` is for cases where the function itself may raise, like this:
```python
def foo(arg=None):
    if arg is None:
        raise Exception
    return arg
```

But what we have is a case where the return value itself is capable of raising an exception when iterated, and that comment is to draw attention to that. Should I make the comment clearer?

lucc replied:
Yea generators are strange in that respect especially if they are not used with the default pattern of 
```py
for x in gen():
    pass
```
but
```py
g = gen()
for x in g:
    pass
```
We can also leave it like it is.

lucc commented on alot/crypto.py:
```diff
@@ -154,111 +152,87 @@ def detached_signature_for(plaintext_str, key=None):
     :param encrypt: whether the key should be able to encrypt
     :type encrypt: bool
-
+    :raises ~alot.errors.GPGProblem: is the key is revoked, expired, or invalid
```
"if" not "is"

dcbaker replied:
oops

lucc commented on tests/crypto_test.py:
```diff
@@ -66,65 +64,90 @@ def tearDownModule():
-        _, detached = crypto.detached_signature_for(to_sign, ctx.get_key(FPR))
+        with gpg.core.Context() as ctx:
+            to_sign = "this is some text.\nit is more than nothing.\n"
+            _, detached = crypto.detached_signature_for(to_sign, ctx.get_key(FPR))
```
I think the with block can end here. The lines below need not be indented.

lucc commented on tests/crypto_test.py:
```diff
@@ -66,65 +64,90 @@ def tearDownModule():
-        _, detached = crypto.detached_signature_for(to_sign, ctx.get_key(FPR))
+        with gpg.core.Context() as ctx:
+            to_sign = "this is some text.\nIt's something\n."
+            _, detached = crypto.detached_signature_for(to_sign, ctx.get_key(FPR))
```
end with block here?

lucc commented on tests/crypto_test.py:
```diff
@@ -66,65 +64,90 @@ def tearDownModule():
+        with gpg.core.Context() as ctx:
+            to_sign = "this is some text.\nIt's something\n."
+            similar = "this is some text.\r\n.It's something\r\n."
+            _, detached = crypto.detached_signature_for(to_sign, ctx.get_key(FPR))
```
end with block here?

lucc commented on tests/crypto_test.py:
```diff
@@ -236,22 +259,32 @@ def test_list_hint(self):
-        actual = crypto.get_key(FPR).uids[0].uid
+        with gpg.core.Context() as ctx:
+            expected = ctx.get_key(FPR).uids[0].uid
+            actual = crypto.get_key(FPR).uids[0].uid
```
actual need not be in the with block

lucc commented on tests/crypto_test.py:
```diff
@@ -236,22 +259,32 @@ def test_list_hint(self):
-        actual = crypto.get_key(FPR, validate=True, encrypt=True, sign=True).uids[0].uid
+        with gpg.core.Context() as ctx:
+            expected = ctx.get_key(FPR).uids[0].uid
+            actual = crypto.get_key(FPR, validate=True, encrypt=True, sign=True).uids[0].uid
```
long with block

lucc commented on .travis.yml:
```diff
@@ -48,21 +48,52 @@ before_install: |
 
+    # Set paths
+    export LD_LIBRARY_PATH=$HOME/.local/lib
+    export PKG_ConFIG_PATH=$HOME/.local/lib/pkgconfig
```
upper case?

dcbaker replied:
yup. although it seems to be working without, which is surprising.

dcbaker commented:
@lucc: I think that the encrypt_by_default thing is working now. Can you have a look?

lucc commented:
No I can still not reply. I think the new type of exception should be handled in alot/commands/utils.py:76

dcbaker commented:
Okay, I replicated your issue (I hope this time!) and have a proper fix (hopefully!). So please give it another go.

lucc commented:
After a quick test it seems to work.

lucc commented on tests/crypto_test.py:
```diff
@@ -274,8 +307,12 @@ def test_signed_only_false(self):
+        # use type to make a dynamic sublcass, since raise wont' accept
+        # anything that isn't an Exception
+        error = type('CustomError', (gpg.errors.GPGMEError, ), {})
+        error.getcode = mock.Mock(return_value=gpg.errors.AMBIGUOUS_NAME)
```
Is there any reason for this rather cryptic code? What would be wrong with a local class:
```py
class CustomError(gpg.errors.GPGMEError):
    getcode = mock.Mock(return_value=gpg.errors.AMBIGUOUS_NAME)
```
Or what do you mean by "uninitialzed" here?

dcbaker replied:
it's less typing than a local class? that's the only reason. I can change it

The problem (and the reason that those two tests failed), is that we were mocking over the original class, and that mock wasn't being unset. I'll replace it with a local class which should be clearer.

lucc replied:
I did not know the three arguments form of `type()` and was expecting some deep magic because I was suspecting that there is some kind of thing that is an "unitialized class" that can not be created with the `class` keywoed but only with `type()`.

I vote for a "normal" class definition bacause it comunicates what is happening more obviously. (Maybe I am the only one who didn't know `type()` all that well).

dcbaker replied:
No, I doubt many people know that signature for `type`. `type` is a bit of a mysterious beast in python, it does some really cool things though, like functional class creation (like in this instance), it's also the base class for metaclasses, which are also pretty interesting.

However, unittests really aren't the place to do interesting things, so I've replaced it with a class statement.

lucc commented:
I was just having a long look at `alot.crypto.get_key()` and searching for a way to simplify the long `try-if-except-else-whatnot-try-allover` mess. But I did not find one :/

I had a look at the diff several times since you opend the PR and it looks sane. Now I will just use this branch for a while and see if something unexpected happens. (So this is already a "half approving" review, "half" because the changes are so complex that I don't trust myself to get it all by just looking at the diff.)

pazz commented:
I will also use this as a mixin for a while. Should we set a deadline at which we merge in case nothing terrible happens until then?

dcbaker commented:
I think so. The debian guys are getting a bit antsy to get rid of pygpgme, and they've suggested they'd be willing to just backport this series (or part of it) to 0.5 if they have to. Although I really think we should start thinking about a 0.6 release after this lands.

pazz commented:
agreed

lucc commented:
I was just composing a dummy mail to test some other stuff and addressed it to "hans" (without any "@"). I get this stacktrace:
```
DEBUG:ui:cmdline: 'compose'
DEBUG:ui:bufferlist command string: "compose"
DEBUG:__init__:mode:bufferlist got commandline "compose"
DEBUG:__init__:ARGS: [u'compose']
DEBUG:__init__:cmd parms {'spawn': None, 'sender': None, 'attach': None, 'cc': None, 'omit_signature': False, 'rest': [], 'bcc': None, 'to': None, 'template': None, 'subject': None}
DEBUG:envelope:TEMPLATE: None
DEBUG:ui:Got key (['enter'], [10])
DEBUG:globals:True
DEBUG:globals:[<alot.addressbook.external.ExternalAddressbook object at 0x7f7d150efe50>]
DEBUG:ui:Got key (['h'], [104])
DEBUG:ui:Got key (['a'], [97])
DEBUG:ui:Got key (['n'], [110])
DEBUG:ui:Got key (['s'], [115])
DEBUG:ui:Got key (['enter'], [10])
DEBUG:ui:Got key (['h'], [104])
DEBUG:ui:Got key (['a'], [97])
DEBUG:ui:Got key (['n'], [110])
DEBUG:ui:Got key (['s'], [115])
DEBUG:ui:Got key (['enter'], [10])
DEBUG:globals:SUBJECT: "hans"
DEBUG:globals:Trying to encrypt message because account.encrypt_by_default=trusted
DEBUG:utils:encryption keys: [u'hans']
ERROR:ui:Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1532, in unwindGenerator
    return _inlineCallbacks(None, gen, Deferred())
  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File "/home/luc/vcs/alot/alot/commands/utils.py", line 45, in set_encrypt
    signed_only=signed_only)
  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1532, in unwindGenerator
    return _inlineCallbacks(None, gen, Deferred())
--- <exception caught here> ---
  File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File "/home/luc/vcs/alot/alot/commands/utils.py", line 75, in _get_keys
    signed_only=signed_only)
  File "/home/luc/vcs/alot/alot/crypto.py", line 119, in get_key
    raise e
gpg.errors.GPGMEError: GPGME: Invalid value
```

lucc commented:
I think this extra test hits the same bug and can be used:
```diff
diff --git a/tests/crypto_test.py b/tests/crypto_test.py
index 048323e3..3d460fda 100644
--- a/tests/crypto_test.py
+++ b/tests/crypto_test.py
@@ -294,0 +295,5 @@ class TestGetKey(unittest.TestCase):
+    def test_missing_key_plain_search_string(self):
+        with self.assertRaises(GPGProblem) as caught:
+            crypto.get_key('foobar')
+        self.assertEqual(caught.exception.code, GPGCode.NOT_FOUND)
+
```

lucc commented:
I just confirmed that this does not happen on master. You can try with `alot compose --sender your-configured-address@alot-config.file --to foobar --subject foo`.

dcbaker commented:
Yeah, we had a test like that but I changed it to cover the MissingKey bug and forgot to copy it. I'll push an updated.

jljusten commented:
Ah. Good to see that this is getting close to being merged now. Thanks @dcbaker!

It looks like #1082 is also a blocker for the [0.6](https://github.com/pazz/alot/milestone/8) release. Is it close as well?

lucc commented:
I was just getting lost with all these gpg transition issues and PRs so:
Fixes #1069

dcbaker commented:
Since things are starting to look settled, I've rebased on master and updated the commit message of the main patch to replace all of the "v<x>" changsets with a more descriptive output, and added a "fixes" tag.

pazz commented on tests/crypto_test.py:
```diff
@@ -66,29 +64,54 @@ def tearDownModule():
+def make_key(revoked=False, expired=False, invalid=False, can_encrypt=True,
+             can_sign=True):
+    # This is ugly
+    mock_key = mock.create_autospec(gpg._gpgme._gpgme_key)
```
Just spotted that this line mentions gpgme still. Is this accidental?

lucc replied:
That is correct and part of the "new" bindings. Try `python2 -c 'import gpg; print gpg._gpgme; print gpg._gpgme._gpgme_key'` in the shell.

pazz replied:
thanks for clarifying this.

pazz commented:
Thanks for all the hard work on this one everyone!

lucc commented:
Horray!! Thanks @dcbaker especially!
